feat: persistent channel watcher, request peer to close channel

kincaidoneil opened the pull request:
- channel watcher runs even if peer isn't connected
- internal refactoring to enable accounts to exist in the background
- load all stored accounts on connect (e.g. if server restarts, it starts the channel watcher for every account)
- performance improvements when signing/validating claims
- clients can send a `requestClose` subprotocol message, triggering the peer to claim the channel, if it's profitable

Currently, there's no garbage collection for idle accounts when the peer isn't connected and the channel watcher isn't running/has no open channel to watch.

kincaidoneil committed:
feat: simplify server and client wrapper

- fetch/load account via functions passed to constructor
- separation of concerns: accounts are no longer instantiated in plugin wrapper
- add specific types for mini-accounts and btp options

kincaidoneil committed:
chore: upgrade btp-packet, use eventemitter3

- add .env to gitignore
- use esModuleInterop in tsconfig

kincaidoneil committed:
feat: top-level accounts in master

- inject data and money handlers to each account, register on master
- load all accounts on connect
- unload all accounts on disconnect
- account data is loaded from store as new peers connect
- rename claimOnDisconnect to closeOnDisconnect
- update imports to work with esModuleInterop
- default settling periods change based on client/server
- cache network and contract as optimization
- add mini-accounts/btp options to exported plugin options

kincaidoneil committed:
feat: single priority queue for all settlements

- including when the channel watcher runs

kincaidoneil committed:
feat: cache channel state, network and contract

- create payment digest locally, not in contract
- should mostly eliminate remote eth calls when sending & validating claims
- simplify event emitter for pending tx

kincaidoneil committed:
chore: update deps, btp-packet

kincaidoneil committed:
feat: inject account data into constructor

- update account to support top level accounts in master, rather than plugin wrapper
- when master loads the account, it fetches the saved data from the store, and injects it into constructor
- after EthereumAccount is instantiated, it's ready to go, rather than waiting for _connect
- remove fetching of channel state in connect: that can happen on-demand

kincaidoneil committed:
fix: more useful logs during settlement, reorg queue

- attemptSettle is queued, rather than each invocation of attemptSettle
- only log when settlement is explicitly triggered for clarity

kincaidoneil committed:
fix: simplify should fund channel logic

kincaidoneil committed:
feat: add requestClose subprotocol for peer to close channel

- peer will still only close if it's profitable
- the incoming channel fee should be used to account for the claim tx fee

kincaidoneil committed:
feat: add persistent channel watcher (#20)

- starts automatically when account is instantiated
- since all accounts are loaded initially, channel watcher works even if server restarts
- add more robust test case where server restarts
- if channel watcher isn't running and a channel is linked/new incoming claim, channel watcher starts
- if channel is closed or no claim exists, channel watcher stops itself
- channel watcher is stopped when the account is unloaded

kincaidoneil committed:
test: minor fixes

- compatibility with 'strict' mode
- fix imports for esModuleInterop

emschwartz commented on src/account.ts:
```diff
@@ -36,72 +52,104 @@ export const getSubprotocol = (message: BtpPacket, name: string) =>
-  // Queue of (at most) a single transaction to claim the incoming channel
-  private claimTransaction = new Mutex(1)
+  // Single queue for all incoming settlements, outgoing settlements and channel claims
+  private queue = new Mutex()
```
queue != mutex, it's kind of strange for these to be named the way they are

emschwartz commented on src/account.ts:
```diff
@@ -274,20 +269,19 @@ export default class EthereumAccount {
-    let amountLeftover = new BigNumber(0)
+    return this.queue.synchronize(TaskPriority.Outgoing, async () => {
+      let settlementBudget = new BigNumber(0)
+      let amountLeftover = new BigNumber(0)
```
Why are these declared at the top instead of where they're used?

emschwartz commented on src/account.ts:
```diff
@@ -301,82 +295,86 @@ export default class EthereumAccount {
+        this.account.payoutAmount = this.account.payoutAmount.minus(settlementBudget)
+
+        // Convert gwei to wei
+        settlementBudget = convert(settlementBudget, Unit.Gwei, Unit.Wei).dp(0, BigNumber.ROUND_FLOOR)
```
Probably better for this to have a different name to a) avoid confusion and b) avoid bugs if this code is modified later

emschwartz commented on src/account.ts:
```diff
@@ -274,20 +269,19 @@ export default class EthereumAccount {
+      let amountLeftover = new BigNumber(0)
 
-    try {
       // Don't attempt settlement if there's no configured settle threshold ("receive only" mode)
```
Wouldn't it be clearer if there were just an explicit variable it would check for (like `receiveOnly`)?

kincaidoneil replied:
Thoughts on adding an `isReceiveOnly` getter to master? (to me it just seems a little verbose given that's the only time `settleThreshold` is referenced)

emschwartz commented on src/account.ts:
```diff
@@ -301,82 +295,86 @@ export default class EthereumAccount {
+        const leftover = format(amountLeftover, Unit.Wei)
+        const account = this.account.accountName
+
+        if (amountSettled.eq(settlementBudget)) {
```
It would be helpful to explain with comments why we might end up with each of these results

emschwartz commented on src/account.ts:
```diff
@@ -408,42 +409,35 @@ export default class EthereumAccount {
-
-        let confHandler: (confNumber: number, receipt: TransactionReceipt) => void
-        let errorHandler: (err: Error) => void
+        const emitter = this.master._web3.eth.sendTransaction(tx) as unknown as EventEmitter
```
why the two `as` statements?

kincaidoneil replied:
TypeScript throws an error than the type can't be directly converted to an `EventEmitter` since there's not sufficient overlap, but converting it to `unknown` first fixes it. Ideally @types/web3 would be correct...but alas.

emschwartz commented on src/account.ts:
```diff
@@ -676,45 +675,45 @@ export default class EthereumAccount {
-            && typeof claim.channelId === 'string'
-            && typeof claim.signature === 'string'
-          if (!hasValidSchema) {
+          const hasValidSchema = (o: any): o is IClaim => {
```
Why does this need to be a closure instead of just an if statement?

kincaidoneil replied:
I wasn't able to find a way to get the type guard `o is IClaim` to work correctly unless it was a function -- any ideas? (Otherwise `claim` will just be `any`).

emschwartz commented on src/account.ts:
```diff
@@ -824,109 +811,139 @@ export default class EthereumAccount {
         const claim = this.account.bestIncomingClaim
-        if (!claim) return
+        if (!claim) {
+          // No claim is linked: stop the channel watcher
```
Would there be a claim right when it starts up?

kincaidoneil replied:
Yes, if a claim was persisted in the store for that account (e.g. if the server restarted, the channel watcher starts for each account).

emschwartz commented on src/account.ts:
```diff
@@ -824,109 +811,139 @@ export default class EthereumAccount {
+    )
 
-    // Check if we're in a Node.js environment
+    // Don't let the timer prevent existing the process if we're in Node.js
```
Should it make sure it's done claiming before exiting?

kincaidoneil replied:
Hmmm...I'll get rid of that. It'd be weird to have the process exit while the channel watcher is still running -- that's probably unexpected behavior.

(This also suggests unref is an anti-pattern: https://www.bennadel.com/blog/3301-calling-timeout-unref-with-settimeout-does-not-appear-to-be-a-best-practice-in-node-js.htm).

emschwartz commented on src/account.ts:
```diff
@@ -824,109 +811,139 @@ export default class EthereumAccount {
+      // Check to verify it's profitable first
+      // This fee should already be accounted for in balance as a part of the initiation fee
+      const txFee = new BigNumber(tx.gasPrice).times(tx.gas)
+      if (txFee.gte(claim.value)) {
```
If it's 1 Wei more, should it still claim it?

kincaidoneil replied:
Practically speaking, yeah, a net gain of 1 wei isn't very useful -- and you'd probably rather keep the channel open to save on fees the connector charges you if you want to use it in the future (deposits to existing channels are cheaper than opening new channels).

I would prefer not to add more config options (e.g. `minAmountToClaim`), but it's probably an argument for adding an admin API to close channels, so clients (or servers) can have their own logic for when to do so. That way it'd also be easier to expose to moneyd-gui for manually closing channels. (Do you have any ideas on that?)

emschwartz reviewed (commented):
Put a couple of comments. However, I find the flow of the logic somewhat difficult to follow so I don't feel super confident about my ability to comment on whether it does the right thing in all of the edge cases.

kincaidoneil committed:
chore: client and server to plugin/

kincaidoneil committed:
fix: lint test files

- remove MetaMask api top-up
- fix linting errors
- update to newer version of stream (receive-only branch)

kincaidoneil committed:
fix: address comments

- clarity in settlement logic
- rename mutex to queue
- remove unref

kincaidoneil commented:
@emschwartz Fixed some issues per your comments and tried to make the settlement logic clearer.

emschwartz commented on src/account.ts:
```diff
@@ -255,106 +255,97 @@ export default class EthereumAccount {
+        // 1) Try to send a claim: spend the channel down entirely before funding it
+        this.sendClaim(budget)
+        // 2) Open or deposit to a channel if it's necessary to send anything leftover
+        .then(checkBudget(l => this.fundOutgoingChannel(l)))
```
What is `l`? In general I'd prefer to avoid single-letter variable names (especially `l` because it looks so much like a `1`)

emschwartz reviewed (approved):
Looks okay to me but it would be good to have others take a look as well. @sentientwaffle @sharafian @austin-king @d1no007

karzak commented on src/account.ts:
```diff
@@ -36,72 +52,104 @@ export const getSubprotocol = (message: BtpPacket, name: string) =>
+    this.watcher = this.startChannelWatcher()
+  }
+
+  private async fetchChannel (channelId: string): Promise<IChannel | undefined> {
```
Maybe call this `updateChannelChache` to avoid the double naming and make it clear that this does more than a simple fetch.

karzak commented on src/utils/queue.ts:
```diff
@@ -1,41 +1,55 @@
+
+  /**
+   * Add a new task to the queue with the given priority
+   * @param priority greater numbers represent tasks that will run sooner than lesser numbers
```
I would add a comment about what the expected behavior is for tasks with the same priority. FIFO, LIFO, no guarantee? Not clear to me at least.

kincaidoneil committed:
fix: address comments, stable priority queue

- JS sort() is not a stable sort; different behavior between V8 versions
- switch to p-queue, which has more features (e.g. idle) and stable sort

kincaidoneil committed:
fix: claim on disconnect

kincaidoneil committed:
fix: update mini-accounts, ilp-packet

- remove incoming expiration timer

kincaidoneil committed:
3.0.0-beta.8
